scylla/src/statement/bound.rs

```diff
@@ -0,0 +1,59 @@
+use scylla_cql::serialize::{
+    row::{SerializeRow, SerializedValues},
+    SerializationError,
+};
+
+use crate::routing::Token;
+
+use super::prepared::{
+    PartitionKey, PartitionKeyError, PartitionKeyExtractionError, PreparedStatement,
+};
+
+/// Represents a statement that already had all its values bound
+#[derive(Debug, Clone)]
+pub struct BoundStatement {
+    pub(crate) prepared: PreparedStatement,
+    pub(crate) values: SerializedValues,
+}
+
+impl BoundStatement {
+    pub(crate) fn new(
+        prepared: PreparedStatement,
+        values: &impl SerializeRow,
+    ) -> Result<BoundStatement, SerializationError> {
+        let values = prepared.serialize_values(values)?;
+        Ok(Self { prepared, values })
+    }
+
+    /// Determines which values constitute the partition key and puts them in order.
+    ///
+    /// This is a preparation step necessary for calculating token based on a prepared statement.
+    pub(crate) fn pk(&self) -> Result<PartitionKey<'_>, PartitionKeyExtractionError> {
+        PartitionKey::new(self.prepared.get_prepared_metadata(), &self.values)
+    }
+
+    pub(crate) fn pk_and_token(
+        &self,
+    ) -> Result<Option<(PartitionKey<'_>, Token)>, PartitionKeyError> {
+        if !self.prepared.is_token_aware() {
+            return Ok(None);
+        }
+
+        let partition_key = self.pk()?;
+        let token = partition_key.calculate_token(self.prepared.get_partitioner_name())?;
+        Ok(Some((partition_key, token)))
+    }
+
+    /// Calculates the token for the prepared statement and its bound values
+    ///
+    /// Returns the token that would be computed for executing the provided prepared statement with
+    /// the provided values.
+    pub fn token(&self) -> Result<Option<Token>, PartitionKeyError> {
+        self.pk_and_token().map(|p| p.map(|(_, t)| t))
+    }
+
+    /// Returns the prepared statement behind the `BoundStatement`
+    pub fn prepared(&self) -> &PreparedStatement {
+        &self.prepared
+    }
```

Comment:
:question: The commit message says that the `BoundStatement` is internal-only. At the same time, it's `pub` and it has `pub` methods.

Reply:
There is no public way to create it, the only method `PreparedStatement::bind` is `pub(crate)`. But I am fine making the struct itself `pub(crate)` as well for this commit and then undoing it in the commit that makes it public.